### DISCUSSION TITLE
Helpful Error Messages. We need them.

### DISCUSSION THREAD

**Discussion by reinismx** [scandipwa/scandipwa#2612] [Ideas]
Error messages are great! Let's make them helpful and constructive in Scandi. This will make it easier to catch bugs, making learning faster and the overall developer experience better.

The Rust programming language has a great feature – as soon as the developer makes a mistake, the compiler warns about it. But instead of a brief and confusing error code, it provides a full explanation including:

- What went wrong
- How to fix it
- Link to a page for more information (the Rust [error index](https://doc.rust-lang.org/error-index.html#E0415))

Let's steal this.

### ESLint rules
Suppose a developer overrides the Header component and names it `Header`.

Error:

```
src/component/Header/Header.component.js
  Line 13:14:  Class name must be derived from the file name, using postfix  @scandipwa/scandipwa-guidelines/derived-class-names
```

This is confusing for anyone that doesn't already know what the issue is. Instead, they should see a helpful error message:

```
src/component/Header/Header.component.js
  Line 13:14:  @scandipwa/scandipwa-guidelines/derived-class-names
    In Scandi, by convention, class names need to be based on the file name.
    Help: To fix this error, the class should be named HeaderComponent, since the file name is Header.component.js
    To learn more, visit <helpful documentation link URL>
```

Searching the Slack channel for the current error message, we can see that it has already caused a lot of confusion: [example](https://scandipwa.slack.com/archives/CD8ETGY3Y/p1611820112208100), [example](https://scandipwa.slack.com/archives/CD8ETGY3Y/p1619000981288500), [example](https://scandipwa.slack.com/archives/CD8ETGY3Y/p1611730834168100), [example](https://scandipwa.slack.com/archives/CD8ETGY3Y/p1618832783224700) (there are more!).

Note: I'm suggesting we do this for all custom ESLint error messages, not just this one. 

### scandipwa CLI
Currently, if the developer mistypes a Scandi CLI command, the CLI terminates quietly.

For example:
```
# should be `extension create`, but easy to make this mistake given that `create component` is valid
scandipwa create extension my-extension

# typo in component
scandipwa create compnent SomeComponent
```

Both of these cause the CLI to terminate without any output. The developer is left wondering... did the command succeed? was there an error?

Instead, there should be an error message:
```
Error: no such command, create compnent
```

And ideally, a helpful suggestion:

```
Help: Did you mean "create component"?
For more information, read the <link to documentation>
```

### Plugin System
Suppose someone is writing a plugin for the `render` method but uses `member-property` instead of `member-function` by mistake:
```
export default {
    "Component/App/Component": {
        "member-property": {
            render: (args, callback, instance) => <>Hello world{callback(...args)}</>
        }
    }
}
```

They would get this very confusing error message:
```
TypeError: args is not iterable
render
src/plugin/App.plugin.js:4

  1 | export default {
  2 |     "Component/App/Component": {
  3 |         "member-property": {
> 4 |             render: (args, callback, instance) => <>Hello world{callback(...args)}</>
  5 |         }
  6 |     }
  7 | }
```

Instead, this mistake should be caught earlier, and they should get a helpful error message:

```
Error: `render` is a member function in `Component/App/Component`, but you tried to use a `member-property` plugin for it.
Help: Did you mean to use a `member-function` plugin?
For more information, check the <link to plugin documentation>
```

## tldr
Scandi should catch all errors as soon as possible, and provide helpful suggestions as well as links to documentation resources. By making errors more helpful, Scandi can accelerate the developer on-boarding process, as well as improve the general developer experience.

**Comment by ejnshtein** [comment]
Totally agree.

Also, if you are working in Magento mode, when upgrading the scandipwa version, essentially `npm i @scandipwa/scandipwa@latest` you also need to upgrade dependencies in `composer.json`.

But what really frustrates me is that for each wrong version in `composer.json` you are shown only **one** error message at the time, so if you need to update 10 dependencies at the time, **you literally need to run one command 10 times** or go to changelog which is not very convenient to do from the command line.

![Screenshot_20210608_155828](https://user-images.githubusercontent.com/18352350/121189032-60c0b100-c872-11eb-8f3c-781c47f0ea7c.png)


Ideally will be to show what is wrong **a single time** and maybe make a prompt to automatically bump those incorrect dependencies versions to correct ones.

To summarize - **What can be done automatically - should be done automatically**
